Source/Utils/CoarsenMR.H

```diff
@@ -0,0 +1,99 @@
+                  const int j,
+                  const int k,
+                  const int comp )
+    {
+        // Indices of destination array (coarse)
+        const int ic[3] = { i, j, k };
+
+        // Number of points and starting indices of source array (fine)
+        int np[3], idx_min[3];
+
+        // Compute number of points
+        for ( int l = 0; l < 3; ++l ) {
+            if ( cr[l] == 1 ) np[l] = 1; // no coarsening
+            else              np[l] = cr[l]*(1-sf[l])*(1-sc[l])     // cell-centered
+                                      +(2*(cr[l]-1)+1)*sf[l]*sc[l]; // nodal
+        }
+
+        // Compute starting indices of source array (fine)
+        for ( int l = 0; l < 3; ++l ) {
+            if ( cr[l] == 1 ) idx_min[l] = ic[l]; // no coarsening
+            else              idx_min[l] = ic[l]*cr[l]*(1-sf[l])*(1-sc[l])     // cell-centered
+                                           +(ic[l]*cr[l]-cr[l]+1)*sf[l]*sc[l]; // nodal
+        }
+
+        // Auxiliary integer variables
+        const int numx = np[0];
+        const int numy = np[1];
+        const int numz = np[2];
+        const int imin = idx_min[0];
+        const int jmin = idx_min[1];
+        const int kmin = idx_min[2];
+        const int sfx  = sf[0];
+        const int sfy  = sf[1];
+        const int sfz  = sf[2];
+        const int scx  = sc[0];
+        const int scy  = sc[1];
+        const int scz  = sc[2];
+        const int crx  = cr[0];
+        const int cry  = cr[1];
+        const int crz  = cr[2];
+        int  ii, jj, kk;
+        Real wx, wy, wz;
+
+        // Add neutral elements (=0) beyond guard cells in source array (fine)
+        auto const arr_src_safe = [arr_src]
+            AMREX_GPU_DEVICE (int const ix, int const iy, int const iz, int const n) noexcept
+            {
+                return arr_src.contains( ix, iy, iz ) ? arr_src(ix,iy,iz,n) :  0.0_rt;
+            };
+
+        // Interpolate over points computed above
+        Real c = 0.0_rt;
+        for         (int kref = 0; kref < numz; ++kref) {
+            for     (int jref = 0; jref < numy; ++jref) {
+                for (int iref = 0; iref < numx; ++iref) {
+                    ii = imin+iref;
+                    jj = jmin+jref;
+                    kk = kmin+kref;
+                    wx = (1.0_rt/static_cast<Real>(numx))*(1-sfx)*(1-scx) // if cell-centered
+                         +((abs(crx-abs(ii-i*crx)))/static_cast<Real>(crx*crx))*sfx*scx; // if nodal
+                    wy = (1.0_rt/static_cast<Real>(numy))*(1-sfy)*(1-scy) // if cell-centered
+                         +((abs(cry-abs(jj-j*cry)))/static_cast<Real>(cry*cry))*sfy*scy; // if nodal
+                    wz = (1.0_rt/static_cast<Real>(numz))*(1-sfz)*(1-scz) // if cell-centered
+                         +((abs(crz-abs(kk-k*crz)))/static_cast<Real>(crz*crz))*sfz*scz; // if nodal
+                    c += wx*wy*wz*arr_src_safe(ii,jj,kk,comp);
+                }
+            }
+        }
```

Comment:
As this is not very easy, could you explain what you are doing here? It doesn't have to be long, something like 2 lines, but just to explain roughly how many cells we loop over, what type of averaging stencil we use and why we do that.